# **Tarea 2 Módulo 3**

## Alumno: Nicolás Bene

---

Vamos a analizar el perfil de Rosalía (@rosalia) en Twitter. 

Las tareas que debes realizar y las preguntas a las que debes dar respuesta en esta actividad son la siguientes:

1. Autentícate en twitter. Pon en notebook todo el código necesario y deja con “XXXXXX” las credenciales (bearer_token) para que el profesor se pueda autenticar con sus propias credenciales y ejecutar el resto del notebook. 
1. ¿Cuál es el id de @rosalia en twitter?
1. Lista los nombres de los usuarios a los que @rosalia está siguiendo.
1. ¿Cuál de ellos tiene más seguidores?
1. Utiliza la get_users_tweets() para obtener el json completo con los cinco últimos tweets de Rosalia. Muestra por pantalla el día, la hora de creación y el texto.
1. Recupera los cinco primeros tweets, a partir de la SteamingClient, desde de la entidad del país/región/ciudad de tu elección y que no sean retweets. Muestra por pantalla el id del usuario que realiza el tweet, la fecha, la hora de creación y el texto. 

<br>

Primero importo las librerías

In [1]:
import json
import tweepy
import urllib
import time
import pandas as pd

## 1) Autentícate en twitter. Pon en notebook todo el código necesario y deja con “XXXXXX” las credenciales (bearer_token) para que el profesor se pueda autenticar con sus propias credenciales y ejecutar el resto del notebook. 

Creo las credenciales para posteriormente autenticarme y conectarme a la API de Twitter. Para ello primero creo una variable con mi bearer token.

In [24]:
bearer_token = 'AAAAAAAAAAAAAAAAAAAAAGhfZwEAAAAApP%2BYZkOjG7iDq1qiA9clljU5O3A%3DdWntbbi64GvVGo6oZe1VU1tvowTORQdie9AsexHwsDq60rqyYe'

A continuación me autentico.

In [3]:
client = tweepy.Client(bearer_token=bearer_token, wait_on_rate_limit=True)

## 2) ¿Cuál es el id de @rosalia en twitter?

Para ver el id del usuario solicitado, primero creo una variable con el nombre de la usuaria y obtengo sus datos a través de la API, especificando los `user_fields`.

In [4]:
USER_NAME = "rosalia"

#Obtengo los datos del usuario rosalia
user = client.get_user(username=USER_NAME, 
                       user_fields=["created_at","description","entities","id","location","name","pinned_tweet_id"
                       ,"profile_image_url","protected","public_metrics","url","username","verified","withheld"])

Obtenido así el usuario, me fijo el id analizando dicho atributo del objeto user.

In [22]:
user.data.id

1728956976

El id de @rosalia es, entonces, 1728956976.

## 3) Lista los nombres de los usuarios a los que @rosalia está siguiendo.

Primero me fijo la cantidad de usuarios a los que @rosalia sigue.

In [6]:
user.data.public_metrics["following_count"]

54

Son 54, por lo tanto se pueden listarlos fácilmente por pantalla.

Para verlos uso el comando `get_users_following`, ya que el mismo muestra las personas a las que sigue @rosalia. Para listarlos a todos  se puede utilizar un objeto `Paginator`, y dentro de cada página me quedo con el atributo `user_name` que es el nombre del usuario Rosalia sigue. Cada nombre de usuario lo voy asignando a una lista creada previamente, llamada following.

Posteriormente controlaré que la cantidad de elementos de esta lista coincida con el conteo realizado anteriormente.

In [7]:
#Creo la lista en la que luego almacenaré los nombres de usuarios
following = []

#Realizo el loop con el que se llenará la lista anterior
for page in tweepy.Paginator(client.get_users_following, user.data.id, max_results=100, limit = 14):
  for response in page.data:
    following.append(response.data["username"])

El loop anterior creó la lista following, donde se guardó el nombre de los usuarios a los que sige @rosalia. Me fijo el número de elementos de dicha lista para cerciorarme que da igual que el `user.data.public_metrics["following_count"]`.

In [23]:
len(following)

54

El número de elementos es igual que el conteo que se hizo. Resta listar los nombres de los 54 usuarios a los que Rosalia sigue:

In [9]:
for i in following:
    print(i)

NoeliaRMontes
jose272011
neggata
jooleesah
rociorotten
caesarenas
katbeee
motovenus
zanelowe
dantijerina
RollingStoneUK
carolineplz
ivangb00
iamcardib
Tainy
rossydpalma
officialDannyT
DONOMAR
OTHERtone
brokemogul
ChadHugo
Pharrell
TheNeptunes
babykeem
Harry_Styles
tokischa_
secoweedcodein
kaydyCain
dragon___loco
caetanoveloso
0PN
jimmyfallon
MuraiHiro
MayaCPopa
historiayarte_
yokoono
elonmusk
rauwalejandro
MissyElliott
YouTubeJapan
jamesblake
CharmLadonna
arca1000000
Camaron_Fans
ptobella
playboicarti
caroshawmusic
KylieJenner
ChloeBailey
rojuuaka
RodoredaMerce
leticia__sala
michaeluzowuru
daikyri_


## 4) ¿Cuál de ellos tiene más seguidores?

Para contestar a esta pregunta, lo primero que hago es un loop para crear una lista con los followers de cada uno de los 54 usuarios a los que sigue Rosalia. 

In [10]:
#Creo la lista en la que luego almacenaré el conteo de los followers de los usuarios a los que sigue rosalia
f_counts = []

#Realizo el loop que llena la lista anterior
for i in following:
    f_counts.append(
        client.get_user(username=i, 
                        user_fields=["public_metrics"])
            .data.public_metrics["followers_count"]
        
    )
    

Luego paso las dos listas creadas (la de usuarios y la del conteo de sus followeres) a un dataframe.

In [11]:
#Creo el dataframe
following_df = pd.DataFrame({"user" : following, "followers_counts": f_counts})

#Lo observo
following_df

user  followers_counts
0    NoeliaRMontes              3553
1       jose272011             39010
2          neggata             20800
3        jooleesah              3440
4      rociorotten              1473
5       caesarenas              1584
6          katbeee              4181
7        motovenus                45
8         zanelowe            676661
9      dantijerina               344
10  RollingStoneUK              3470
11     carolineplz             81761
12        ivangb00               555
13       iamcardib          22109327
14           Tainy            105196
15     rossydpalma             67913
16  officialDannyT            469415
17         DONOMAR           5279608
18       OTHERtone              1853
19      brokemogul             60679
20        ChadHugo             35112
21        Pharrell          10659820
22     TheNeptunes             19224
23        babykeem            206254
24    Harry_Styles          37433179
25       tokischa_            121669
26  secoweedcodein            160590
27       kaydyCain            135276
28   dragon___loco               280
29   caetanoveloso           1026985
30             0PN             67634
31     jimmyfallon          51455301
32       MuraiHiro             31186
33       MayaCPopa             21132
34  historiayarte_             96594
35         yokoono           4609290
36        elonmusk          77511081
37   rauwalejandro           1842271
38    MissyElliott           6966922
39    YouTubeJapan            414164
40      jamesblake            339648
41    CharmLadonna              7454
42     arca1000000            118724
43    Camaron_Fans             24275
44        ptobella              1177
45    playboicarti           2230208
46   caroshawmusic              8072
47     KylieJenner          39696114
48     ChloeBailey           1147835
49        rojuuaka             76436
50   RodoredaMerce             31285
51   leticia__sala              6372
52  michaeluzowuru             10018
53        daikyri_             34778

Me fijo ahora cuál usuario de los que Rosalia sigue tiene más seguidores.

In [12]:
following_df[following_df.followers_counts == following_df.followers_counts.max()]

user  followers_counts
36  elonmusk          77511081

Dentro de los 54 twitteros, Elon Musk es el que tiene más seguidores, con más de 77 millones.

## 5) Utiliza la get_users_tweets() para obtener el json completo con los cinco últimos tweets de Rosalia. Muestra por pantalla el día, la hora de creación y el texto. ##

In [15]:
#Busco los últimos 5 tweets de Rosalia
tweets = client.get_users_tweets(id=user.data.id, tweet_fields=["created_at"],max_results =5)

#Imprimo los datos solicitados de los tweets
for tweet in tweets.data:
    print(tweet.created_at, tweet.text)

2022-03-12 04:38:14+00:00 RT @johandls360: NO PUEDO ESPERARRRRrrrrrr @rosalia #MOTOMAMI https://t.co/rN1aC2ibc3
2022-03-12 00:06:26+00:00 @tokischa_ @jimmyfallon Eso tUUUUU todo el rato toKi💓
2022-03-11 21:07:02+00:00 Aunque
14/20:

El casco es el mejor bolso de unA motomami
2022-03-11 21:06:08+00:00 Si el bolso no combina, mejor no lleves bolso https://t.co/TWsWRm6HJ4
2022-03-11 20:40:27+00:00 FRONTEA CON MOTO, PROTEGE TU MAMI 13/20


## 6) Recupera los cinco primeros tweets, a partir de la SteamingClient, desde de la entidad del país/región/ciudad de tu elección y que no sean retweets. Muestra por pantalla el id del usuario que realiza el tweet, la fecha, la hora de creación y el texto.

In [27]:
# Declaro la clase listener que va a instanciar tweepy.StreamingClient 
class listener(tweepy.StreamingClient):
    # Creo un contador de tweets capturados
    def on_connect(self):
        self.counter = 0
        self.limit = 5
    # Evento que se ejecuta con cada tweet
    def on_tweet(self, tweet):
        print(self.counter)
        print(tweet.id)
        print(tweet.text)
        print(tweet.created_at)
        self.counter += 1
        if self.counter < self.limit:
            return True
        else:
            self.disconnect()
    def on_error(self, tweet):
        print(tweet)

# Declaro la instancia de tweepy.StreamingClient
listen = listener(bearer_token,wait_on_rate_limit=True)

# Elimino las reglas previas
printer_rules = listen.get_rules()
if printer_rules.data is not None:
  listen.delete_rules([str(i.id) for i in listen.get_rules().data])

#Declaro las nuevas reglas, donde mostraré tweets de Uruguay que no sean retweets
listen.add_rules(tweepy.StreamRule('entity:"Uruguay" -is:retweet'))
# Filtro e indico los campos a recuperar. No menciono explicitamente el id ni el texto porque lo muestra por default.
listen.filter(tweet_fields=["created_at"])

0
1502745690178523138
@GuimonacoOk Se te achicó el mapa, se te cayeron Uruguay, Paraguay, Colombia y Ecuador...
2022-03-12 20:37:46+00:00
1
1502745789235355652
it's named after a town in Uruguay where they processed South American meat to make pies. Stop blaming "the Brits" for processed food sold in major urban areas by multinational foodstuffs corporations https://t.co/rCf0eyQT27
2022-03-12 20:38:10+00:00
2
1502745914288574465
@TzaTza416 Mongolia
Italia
Guinea 
Uruguay
El Salvador 👊
Lituania.
2022-03-12 20:38:40+00:00
3
1502746003174313992
Rumanía
Uruguay
Barcelona
ESPAÑA
Niger https://t.co/VCt0oq4kFe
2022-03-12 20:39:01+00:00


Stream connection closed by Twitter


4
1502746019053948928
Un escritor entre Uruguay y Grecia -&gt; https://t.co/r27CIHvKE5 #Berisso
2022-03-12 20:39:05+00:00
